tskv/src/wal.rs

```diff
@@ -1,101 +1,350 @@
-use std::path::{Path, PathBuf};
-
+use std::{
+    any::Any,
+    marker::PhantomData,
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+};
+
+use crc32fast;
+use futures::channel::oneshot;
 use lazy_static::lazy_static;
 use regex::Regex;
-use snafu::Snafu;
+use snafu::prelude::*;
+use tokio::sync::Mutex as AsyncMutex;
 
-use crate::direct_io::{File, FileSync, FileSystem, Options};
-use protos::models::*;
+use crate::{
+    direct_io::{make_io_task, File, FileCursor, FileSync, FileSystem, Options, TaskType},
+    file_manager,
+};
+use protos::models;
+use walkdir::IntoIter;
 
-use crate::FileManager;
+use crate::{option, FileManager};
 
 lazy_static! {
     static ref WAL_FILE_NAME_PATTERN: Regex = Regex::new("_.*\\.wal").unwrap();
 }
 
 const SEGMENT_SIZE: u64 = 1073741824; // 1 GiB
 
+pub enum WalTask {
+    Write {
+        rows: Vec<u8>,
+        cb: oneshot::Sender<WalResult<()>>,
+    },
+}
+
+// pub struct WalScheduler {
+//     sender: Sender<WalTask>,
+// }
+
+// impl WalScheduler {
+//     pub async fn write(&mut self, rows: Vec<u8>) -> WalResult<()> {
+//         let (cb, rx) = oneshot::channel::<WalResult<()>>();
+//         let task = WalTask::Write { rows, cb };
+//         self.sender.send(task).await;
+
+//         rx.await.map_err(|_| WalError::FailedWithChannelReceive)?
+//     }
+// }
+
 #[derive(Snafu, Debug)]
-pub enum Error {
-    #[snafu(display("Unable to walk dir: {}", source))]
+pub enum WalError {
+    #[snafu(display("Unable to walk dir : {}", source))]
     UnableToWalkDir { source: walkdir::Error },
 
     #[snafu(display("File {} has wrong name format to have an id", file_name))]
     InvalidFileName { file_name: String },
 
-    #[snafu(display("Error with file: {}", source))]
-    FileManagerError { source: super::file_manager::Error },
+    #[snafu(display("Error with file : {}", source))]
+    FailedWithFileManager {
+        source: super::file_manager::FileError,
+    },
+
+    #[snafu(display("Error with std::io : {}", source))]
+    FailedWithStdIO { source: std::io::Error },
+
+    #[snafu(display("Error with receiving from channel"))]
+    FailedWithChannelReceive,
+
+    #[snafu(display("Error with sending from channel"))]
+    FailedWithChannelSend,
+
+    #[snafu(display("Error with IO task"))]
+    FailedWithIoTask,
+}
+
+pub type WalResult<T> = std::result::Result<T, WalError>;
+
+pub enum WalEntryType {
+    Write = 1,
+    Delete = 2,
```

Comment:
how deal with delete table & delete series

Reply:
- delete table ,or only some of series, is `Delete`
- delete them with time range, then `DeleteRange`

Reply:
- delete table, or some of series, use `Delete`,
- delete them with a time range, use `DeleteRange`